File: lib/resolver.ts
/**
 * Client-side Hybrid Source Resolver.
 * Mirrors agent/resolver.py logic for offline/SSR use.
 * Rules §3: local always wins on ties.
 */

import type { LatestSource } from '@/types/repo';

export interface ResolverInput {
  file_mtime?: string | null;
  local_commit_time?: string | null;
  remote_push_time?: string | null;
}

export interface ResolverResult {
  source: LatestSource;
  timestamp: Date | null;
}

function parseDate(s?: string | null): Date | null {
  if (!s) return null;
  const d = new Date(s);
  return isNaN(d.getTime()) ? null : d;
}

export function resolveLatestSource(input: ResolverInput): ResolverResult {
  const candidates: Array<{ source: LatestSource; ts: Date | null }> = [
    { source: 'local_filesystem', ts: parseDate(input.file_mtime) },
    { source: 'local_git', ts: parseDate(input.local_commit_time) },
    { source: 'github_remote', ts: parseDate(input.remote_push_time) },
  ];

  const valid = candidates.filter(c => c.ts !== null) as Array<{
    source: LatestSource;
    ts: Date;
  }>;

  if (!valid.length) return { source: 'github_remote', timestamp: null };

  // Max by timestamp; first entry wins on tie (local priority)
  const best = valid.reduce((a, b) => (b.ts > a.ts ? b : a));
  return { source: best.source, timestamp: best.ts };
}

/** Format a source name for display. */
export function formatSource(source: LatestSource): string {
  const map: Record<LatestSource, string> = {
    local_filesystem: 'Local Files',
    local_git: 'Local Git',
    github_remote: 'GitHub',
  };
  return map[source] ?? source;
}

/** Returns relative time string like "3 days ago". */
export function timeAgo(isoString?: string | null): string {
  if (!isoString) return 'Never';
  const d = new Date(isoString);
  if (isNaN(d.getTime())) return 'Unknown';
  const seconds = Math.floor((Date.now() - d.getTime()) / 1000);
  if (seconds < 60) return 'Just now';
  if (seconds < 3600) return `${Math.floor(seconds / 60)}m ago`;
  if (seconds < 86400) return `${Math.floor(seconds / 3600)}h ago`;
  if (seconds < 7 * 86400) return `${Math.floor(seconds / 86400)}d ago`;
  if (seconds < 30 * 86400) return `${Math.floor(seconds / 604800)}w ago`;
  return d.toLocaleDateString('en-US', { month: 'short', day: 'numeric', year: 'numeric' });
}

File: agent/resolver.py
"""Hybrid Source Resolver™.

Determines the true latest activity for a repo by comparing:
  1. local file modification timestamp  (fastest)
  2. local git commit timestamp
  3. remote git push timestamp

Returns the freshest timestamp and names its source.
Never assumes GitHub is authoritative (rules §3).
"""

from __future__ import annotations
from datetime import datetime, timezone
from typing import Optional

from agent.models import LatestSource


def _utc(dt: Optional[datetime]) -> Optional[datetime]:
    """Normalise datetime to UTC-aware."""
    if dt is None:
        return None
    if dt.tzinfo is None:
        return dt.replace(tzinfo=timezone.utc)
    return dt.astimezone(timezone.utc)


def resolve_latest_source(
    file_mtime: Optional[datetime],
    local_commit_time: Optional[datetime],
    remote_push_time: Optional[datetime],
) -> tuple[LatestSource, Optional[datetime]]:
    """
    Compare all three timestamps and return (source_name, timestamp).

    Priority order when timestamps are equal: local_filesystem wins,
    then local_git, then github_remote (local always wins on ties).
    """
    candidates: list[tuple[LatestSource, Optional[datetime]]] = [
        ("local_filesystem", _utc(file_mtime)),
        ("local_git", _utc(local_commit_time)),
        ("github_remote", _utc(remote_push_time)),
    ]

    # Filter out None timestamps
    valid = [(src, ts) for src, ts in candidates if ts is not None]

    if not valid:
        # No information at all — default to github_remote (least surprising)
        return ("github_remote", None)

    # Pick the freshest; on a tie the first in list wins (local preference)
    best_source, best_ts = max(valid, key=lambda x: x[1])  # type: ignore[arg-type]
    return (best_source, best_ts)


def resolve_latest_source_iso(
    file_mtime_iso: Optional[str],
    local_commit_iso: Optional[str],
    remote_push_iso: Optional[str],
) -> tuple[LatestSource, Optional[str]]:
    """Convenience wrapper that accepts ISO strings and returns ISO string."""

    def _parse(s: Optional[str]) -> Optional[datetime]:
        if not s:
            return None
        try:
            return datetime.fromisoformat(s)
        except ValueError:
            return None

    source, ts = resolve_latest_source(
        _parse(file_mtime_iso),
        _parse(local_commit_iso),
        _parse(remote_push_iso),
    )
    return (source, ts.isoformat() if ts else None)
